Add trashed recipe

shingo256 opened the pull request:
### Brief summary of what the package does

Viewing/editing system trash can -- open, view, restore or permanently delete trashed files or directories in trash can with Dired-like look and feel.  The trash can has to be compliant with freedesktop.org spec in <http://freedesktop.org/wiki/Specifications/trash-spec/>

### Direct link to the package repository

https://github.com/shingo256/trashed

### Your association with the package

Maintainer

### Relevant communications with the upstream package maintainer

None needed

### Checklist

Please confirm with `x`:

- [X] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X] My elisp byte-compiles cleanly
- [X] `M-x checkdoc` is happy with my docstrings
- [X] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

shingo256 commented:
I have reopened the pull request.  Please check.

alphapapa commented:
This should probably be implemented using `tabulated-list-mode`.

shingo256 commented:
Thank you for the great advise.  Now I know that and will try to revise my code shortly.

Silex commented:
@shingo256: consider using https://github.com/politza/tablist, it's a wrapper around tabulated-list which make some things easier.

shingo256 commented:
I took a close look at both tabulated-list and tablist and ended up using tabulated-list and revised Trashed with it as I thought it's a better fit.

I realized that tablist is richer and covers some functionalities which I implemented in Trashed.  However, looks like it might be too much for this simple trash can editing, because it's not a build-in package and actually much bigger than Trashed even though it would simplify just a part of Trashed.  And also, since I would need to sort out and add necessary functionalities, the code might get more complicated from another perspective.

I also improved Trashed a little bit with functionalities such as column movement and regexp-based marking inspired by these packages, and now I'm hoping it's good to go.  I also would continue to consider using tablist in future versions at the time when I need to implement more elaborated functionalities.  Any further advice is appreciated and would also appreciate it if you could resume the checking process with this.

manuel-uberti commented:
@shingo256 I am not part of the MELPA team, but FWIW `trashed` looks better now and it works as expected for me. Great job! :)

shingo256 commented:
Thank you!

alphapapa commented:
Great job.  I sent a few more suggestions as an issue on the repo's tracker.

shingo256 commented:
Thank you for the bug reports and the suggestion.  I quickly fixed my code for the 2 issues and looking at your code.  Let's discuss it in the issue threads.

shingo256 commented:
I further improved it to allow user to customize file size and time format.

riscy commented:
This mostly looks great, I just have a couple small comments:

- I believe you shouldn't have to define `trashed-mode-hook` -- it's defined automatically when you `define-derived-mode trashed-mode`, unless you have a specific reason for wanting it to be customizable. 
- It may be going over my head, but I'm having trouble understanding the need for `trashed-load-hook` -- really loading trashed should only happen once, but maybe I'm missing something.

Some minor suggestions:
- trashed.el#L620: It is beneficial to sharp-quote hooks; use `#'`
- trashed.el#L22: Prefer `https` over `http` (if possible)

And a couple minor typos:
- `restration` -> `restoration`, `Restiration` -> `Restoration`
- `acording` -> `according`

shingo256 commented:
Hi,

Thank you for the advice.  

- `trashed-mode-hook`: Since it is declared with `defvar` in `define-derived-mode`, I declared it here with `defcustom` just to make it customizable and show up in the customize list.  I added a comment to explain this, but if you know any better way, please advise.

- `trashed-load-hook`: Yes, it is run just once at the very end of the trashed.el loading.  I just implemented it in the same way as `dired-load-hook` following Dired convention.  User may use it for their own customization such as function override, etc.

If I miss or misunderstand something, please let me know.  I quickly updated the code with your suggestions and corrected typos.  Thank you so much again.

riscy commented:
>  I just implemented it in the same way as `dired-load-hook` following Dired convention.

I just learned something about dired.  Okay, thanks for clarifying. :)

This looks good.  Welcome to MELPA!
